### PR TITLE
Allow aiming anywhere. Mark practice target.

### LINKED ISSUE #21967: Cant safely aim at empty terrain
Because you need to steady your aim to shoot accurately, and you cant steady your aim while aiming at nothing, all shots made with intent of hitting a piece of furniture or terrain are basically random unaimed shots.

Noticed this after almost killing myself with a fusion gun while trying to blow up some reinforced glass.

### LINKED ISSUE #26850: Archery Targets or Pratice Dummies
**Is your feature request related to a problem? Please describe.**  
Practicing on trees or other things causes arrows to land in random locations. 

**Describe the solution you'd like**  
Something you can construct that will allow the player to lock onto it like an enemy and will also collect arrows shot at it. 

**Describe alternatives you've considered**  
Allow arrows to *stick* into trees. 

**Additional context**  
Could be something like the bail of hay or uses pile of hay you get from foraging. Maybe also require some wire or other material to bind it. 
Preferably has a higher chance of preserving arrows fired at it.

### PR DESCRIPTION
**Pull request by OzoneH3**
#### Summary
```SUMMARY: Features "Allow aiming anywhere. Mark practice target."```

#### Purpose of change
Fixes #21967
Fixes #26850

#### Describe the solution
* Allows aiming without a monster being present.
* Will follow monster crossing aim position.
* Mark a practice target via construction menu. (Same concept as mark firewood source)
* Set target position to practice target if no monsters are around.
* Sneaked in a whizz sound for arrows and thonk for bolts.
* Update recoil when changing target via difference in angle changed.

### CODE REVIEW

**Comment by Firestorm01X2** [comment]
Yes! Really!
Thank you!

**Comment by Firestorm01X2** [comment]
Wait a minute. Is it as mean that now player can preaim from the corner while waiting for monster. Excellent. Splendid.

File: src/player.h [hunk 1]
```diff
@@ -1452,6 +1452,8 @@ class player : public Character
         int oxygen;
         int stamina;
         double recoil = MAX_RECOIL;
+        std::weak_ptr<Creature> last_target;
+        tripoint last_target_pos;
```
Comment (BevapDin) [inline src/player.h]: Please wrap this into `cata::optional` instead of using an undocumented sentinel value.

File: src/ranged.cpp [hunk 2]
```diff
@@ -994,7 +1008,7 @@ std::vector<tripoint> target_handler::target_ui( player &pc, target_mode mode,
     ctxt.register_action( "TOGGLE_SNAP_TO_TARGET" );
     ctxt.register_action( "HELP_KEYBINDINGS" );
     ctxt.register_action( "QUIT" );
-    ctxt.register_action( "SWITCH_MODE" );
+    ctxt.register_action( "SWITCH_MODE" ) ;
```
Comment (BevapDin) [inline src/ranged.cpp]: Why this additional space?

File: src/ranged.cpp [hunk 3]
```diff
@@ -1283,8 +1308,7 @@ std::vector<tripoint> target_handler::target_ui( player &pc, target_mode mode,
             if( aim_mode == aim_types.end() ) {
                 aim_mode = aim_types.begin();
             }
-        } else if( ( action == "AIMED_SHOT" || action == "CAREFUL_SHOT" || action == "PRECISE_SHOT" ) &&
-                   target != -1 ) {
+        } else if( ( action == "AIMED_SHOT" || action == "CAREFUL_SHOT" || action == "PRECISE_SHOT" ) ) {
```
Comment (BevapDin) [inline src/ranged.cpp]: Only one pair of parenthesis is needed here.

File: src/savegame.cpp [hunk 4]
```diff
@@ -77,6 +77,9 @@ void game::serialize( std::ostream &fout )
             json.member( "last_target", critter_tracker->temporary_id( *mon ) );
             json.member( "last_target_type", -1 );
         }
+    } else {
+        tripoint templtp = u.last_target_pos ? *u.last_target_pos : tripoint_min;
+        json.member( "last_target_pos", templtp );
```
Comment (BevapDin) [inline src/savegame.cpp]: You can serialize a `cata::optional` directly, no need to convert it into a `tripoint` and serialize that. Same for deserialization.

Comment (OzoneH3) [reply src/savegame.cpp]: I tried to do:
serialize: ```json.member( "last_target_pos", *u.last_target_pos );```
deserialize: ```data.read( "last_target_pos", *u.last_target_pos );```

But this gives me an error on save: ```cataclysm: src/optional.h:41: T &cata:optional<tripoint>::get() [T = tripoint]: Assertion 'full' failed.```

Comment (AMurkin) [reply src/savegame.cpp]: u.last_target_pos is cata::optional. Use it without a *.

Comment (OzoneH3) [reply src/savegame.cpp]: Tried that, but then this happens: (same with data.read)
```src/json.h:583:13: error: no matching member function for call to 'write'
            write( value );
            ^~~~~
src/savegame.cpp:81:14: note: in instantiation of function template specialization 'JsonOut::member<cata::optional<tripoint> >' requested here
        json.member( "last_target_pos", u.last_target_pos );
             ^
```

File: data/json/construction.json [hunk 5]
```diff
@@ -2391,6 +2391,15 @@
         "pre_note" : "Firewood or other flammable materials on a nearby tile marked in this way may be used to automatically refuel fires.  This will be done to maintain light during long-running tasks that require it such as crafting or reading, but not (for example) if you are simply waiting nearby.",
         "pre_special" : "check_no_trap",
         "post_special" : "done_mark_firewood"
+    },{
+        "type" : "construction",
+        "description" : "Mark practive target",
```
Comment (Night-Pryanik) [inline data/json/construction.json]:
```suggestion
        "description" : "Mark practice target",
```

File: data/json/construction.json [hunk 6]
```diff
@@ -2391,6 +2391,15 @@
...
+        "category" : "OTHER",
+        "required_skills" : [],
+        "time" : 0,
+        "pre_note" : "Mark a spot for target practice.  Firing will automatically target the first practice target location ,found in gunrange, if no enemies are around.",
```
Comment (Night-Pryanik) [inline data/json/construction.json]:
```suggestion
        "pre_note" : "Mark a spot for target practice.  Firing will automatically target the first practice target location found in gun range, if no enemies are around.",
```

File: src/savegame.cpp (hunk 4 above)
Comment (OzoneH3) [reply src/savegame.cpp]: @BevapDin Alright found out whats going on here.
You put the de/serialize functions into savegame_json.cpp
I need them in savegame.cpp though. Where'd I best put them
to access from both files without adding any more includes to .h files?

Comment (BevapDin) [reply src/savegame.cpp]: Move those definitions into "enums.h" and make them `inline` (and add a forward declaration of `cata::optional` there as well).

**Comment by OzoneH3** [comment]
>There's a pre-existing issue of the target point being saved with the game instead of with the player.

Done. This way there's no need to move around the de/serialization for cata::optional<tripoint> eigther.

**Pull request merged**